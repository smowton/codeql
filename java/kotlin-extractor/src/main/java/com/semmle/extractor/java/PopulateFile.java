package com.semmle.extractor.java;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import com.github.codeql.Label;
import com.github.codeql.TrapWriter;
import com.github.codeql.KotlinExtractorDbSchemeKt;
import com.semmle.util.trap.pathtransformers.PathTransformer;
import com.semmle.util.files.FileUtil;

// A temporary port of TrapWriter.populateFile et al, until we can just use the existing TrapWriter or split it up

public class PopulateFile {

  private TrapWriter tw;
  private PathTransformer transformer;
  public PopulateFile(TrapWriter tw) {
    this.tw = tw;
    this.transformer = PathTransformer.std();
  }

	private static final String[] keyReplacementMap = new String[127];
	static {
		keyReplacementMap['&'] = "&amp;";
		keyReplacementMap['{'] = "&lbrace;";
		keyReplacementMap['}'] = "&rbrace;";
		keyReplacementMap['"'] = "&quot;";
		keyReplacementMap['@'] = "&commat;";
		keyReplacementMap['#'] = "&num;";
	}

	/**
	 * Escape a string for use in a TRAP key, by replacing special characters with HTML entities.
	 * <p>
	 * The given string cannot contain any sub-keys, as the delimiters <code>{</code> and <code>}</code>
	 * are escaped.
	 * <p>
	 * To construct a key containing both sub-keys and arbitrary input data, escape the individual parts of
	 * the key rather than the key as a whole, for example:
	 * <pre>
	 * "foo;{" + label.toString() + "};" + escapeKey(data)
	 * </pre>
	 */
	public static String escapeKey(String s) {
		StringBuilder sb = null;
		int lastIndex = 0;
		for (int i = 0; i < s.length(); ++i) {
			char ch = s.charAt(i);
			switch (ch) {
			case '&':
			case '{':
			case '}':
			case '"':
			case '@':
			case '#':
				if (sb == null) {
					sb = new StringBuilder();
				}
				sb.append(s, lastIndex, i);
				sb.append(keyReplacementMap[ch]);
				lastIndex = i + 1;
				break;
			}
		}
		if (sb != null) {
			sb.append(s, lastIndex, s.length());
			return sb.toString();
		} else {
			return s;
		}
	}

  public Label populateFile(File absoluteFile) {
    String databasePath = transformer.fileAsDatabaseString(absoluteFile);
    // Ensure the rewritten path is used from now on.
    File normalisedFile = new File(databasePath);
    Label result = tw.getLabelFor("@\"" + escapeKey(databasePath) + ";sourcefile" + "\"");
    KotlinExtractorDbSchemeKt.writeFiles(tw, result, databasePath);
    populateParents(normalisedFile, result);
    return result;
  }

	private Label addFolderTuple(String databasePath) {
		Label result = tw.getLabelFor("@\"" + escapeKey(databasePath) + ";folder" + "\"");
		KotlinExtractorDbSchemeKt.writeFolders(tw, result, databasePath);
		return result;
	}

	/**
	 * Populate the parents of an already-normalised file. The path transformers
	 * and canonicalisation of {@link PathTransformer#fileAsDatabaseString(File)} will not be
	 * re-applied to this, so it should only be called after proper normalisation
	 * has happened. It will fill in all parent folders in the current TRAP file.
	 */
	private void populateParents(File normalisedFile, Label label) {
		File parent = normalisedFile.getParentFile();
		if (parent == null) return;

    Label parentLabel = addFolderTuple(FileUtil.normalisePath(parent.getPath()));
    populateParents(parent, parentLabel);
		KotlinExtractorDbSchemeKt.writeContainerparent(tw, parentLabel, label);
	}

}