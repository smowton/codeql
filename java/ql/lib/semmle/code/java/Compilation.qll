/**
 * Provides a class representing individual compiler invocations that occurred during the build.
 */

import semmle.code.FileSystem

/**
 * An invocation of the compiler. Note that more than one file may be
 * compiled per invocation. For example, this command compiles three
 * source files:
 *
 *   javac Foo.java Bar.java Baz.java
 *
 * Two things happen to each file during a compilation:
 *
 *   1. The file is compiled by a real compiler, such as javac or kotlinc.
 *   2. The file is parsed by the CodeQL front-end.
 *   3. The parsed representation is converted to database tables by
 *      the CodeQL extractor.
 *
 * This class provides CPU and elapsed time information for steps 2 and 3,
 * but not for step 1.
 */
class Compilation extends @compilation {
  /** Gets a textual representation of this element. */
  string toString() {
    exists(string name |
      compilations(this, _, _, name) and
      result = "<compilation " + name + ">"
    )
  }

  /** Gets a file compiled during this invocation. */
  File getAFileCompiled() { result = getFileCompiled(_) }

  /** Gets the `i`th file compiled during this invocation */
  File getFileCompiled(int i) { compilation_compiling_files(this, i, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * front-end.
   */
  float getFrontendCpuSeconds(int i) { compilation_time(this, i, 1, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * front-end.
   */
  float getFrontendElapsedSeconds(int i) { compilation_time(this, i, 2, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * extractor.
   */
  float getExtractorCpuSeconds(int i) { compilation_time(this, i, 3, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * extractor.
   */
  float getExtractorElapsedSeconds(int i) { compilation_time(this, i, 4, result) }

  /**
   * Gets an argument passed to the extractor on this invocation.
   */
  string getAnArgument() { result = getArgument(_) }

  /**
   * Gets the `i`th argument passed to the extractor on this invocation.
   */
  string getArgument(int i) { compilation_args(this, i, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * compiler.
   */
  float getCompilerCpuSeconds() { compilation_compiler_times(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the compiler.
   */
  float getCompilerElapsedSeconds() { compilation_compiler_times(this, _, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * front-end and extractor.
   */
  float getTotalCpuSeconds() { compilation_finished(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the front-end and extractor.
   */
  float getTotalElapsedSeconds() { compilation_finished(this, _, result) }

  /**
   * Holds if this is a compilation of Java code.
   */
  predicate isJava() { this instanceof @javacompilation }

  /**
   * Holds if this is a compilation of Kotlin code.
   */
  predicate isKotlin() { this instanceof @kotlincompilation }

  /**
   * Holds if extraction for the compilation started.
   */
  predicate extractionStarted() { compilation_started(this) }

  /**
   * Holds if the extractor terminated normally. Terminating with an exit
   * code indicating that an error occurred is considered normal
   * termination, but crashing due to something like a segfault is not.
   */
  predicate normalTermination() { compilation_finished(this, _, _) }
}
