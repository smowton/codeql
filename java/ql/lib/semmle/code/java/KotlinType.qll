/**
 * Provides classes and predicates for working with Kotlin types.
 */

import java

class KotlinType extends Element, @kt_type {
}

class KotlinNullableType extends KotlinType, @kt_nullable_type {
  override string toString() {
    exists(ClassOrInterface ci |
           kt_nullable_types(this, ci) and
           result = "Kotlin nullable " + ci.toString())
  }
  override string getAPrimaryQlClass() { result = "KotlinNullableType" }
}

class KotlinNotnullType extends KotlinType, @kt_notnull_type {
  override string toString() {
    exists(ClassOrInterface ci |
           kt_notnull_types(this, ci) and
           result = "Kotlin not-null " + ci.toString())
  }
  override string getAPrimaryQlClass() { result = "KotlinNotnullType" }
}

class KotlinTypeAlias extends Element, @kt_type_alias {
  override string getAPrimaryQlClass() { result = "KotlinTypeAlias" }

  KotlinType getKotlinType() {
    kt_type_alias(this, _, result)
  }
}
