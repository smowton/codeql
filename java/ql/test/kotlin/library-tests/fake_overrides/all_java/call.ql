import java

from MethodAccess c
select c, c.getCallee()
