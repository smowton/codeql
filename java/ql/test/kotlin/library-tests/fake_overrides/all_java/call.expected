| A.java:4:19:4:29 | someFun(...) | OC$C.class:0:0:0:0 | someFun |
